File: backend/keep-alive-wrapper.js
#!/usr/bin/env node

/**
 * Keep-Alive Wrapper for KOL Tracker Backend
 * Ensures the backend service stays running on Render with auto-restart on failures
 */

const { spawn } = require('child_process');
const http = require('http');

// Configuration
const HEALTH_CHECK_INTERVAL = parseInt(process.env.HEALTH_CHECK_INTERVAL) || 300000; // 5 minutes
const RESTART_DELAY = 5000; // 5 seconds
const MAX_RESTART_ATTEMPTS = 10;
const RESTART_WINDOW = 3600000; // 1 hour

let restartCount = 0;
let lastRestartTime = 0;
let backendProcess = null;
let isShuttingDown = false;

// Colors for logging
const colors = {
  reset: '\x1b[0m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m'
};

function log(message, color = 'reset') {
  const timestamp = new Date().toISOString();
  console.log(`${colors[color]}[${timestamp}] ${message}${colors.reset}`);
}

function startBackend() {
  if (isShuttingDown) return;

  log('🚀 Starting KOL Tracker Backend...', 'green');

  backendProcess = spawn('node', ['simple-server.js'], {
    stdio: ['inherit', 'inherit', 'inherit'],
    env: { ...process.env }
  });

  backendProcess.on('error', (error) => {
    log(`❌ Backend process error: ${error.message}`, 'red');
    restartBackend();
  });

  backendProcess.on('exit', (code, signal) => {
    if (isShuttingDown) {
      log('🛑 Backend shutdown complete', 'yellow');
      return;
    }

    if (code === 0) {
      log('✅ Backend exited gracefully', 'green');
    } else {
      log(`💥 Backend exited with code ${code} and signal ${signal}`, 'red');
      restartBackend();
    }
  });

  // Set up health check
  setTimeout(() => {
    if (!isShuttingDown) {
      startHealthCheck();
    }
  }, 30000); // Wait 30 seconds before starting health checks
}

function restartBackend() {
  if (isShuttingDown) return;

  const now = Date.now();
  
  // Reset restart count if enough time has passed
  if (now - lastRestartTime > RESTART_WINDOW) {
    restartCount = 0;
  }

  restartCount++;
  lastRestartTime = now;

  if (restartCount > MAX_RESTART_ATTEMPTS) {
    log(`🚨 Maximum restart attempts (${MAX_RESTART_ATTEMPTS}) reached within 1 hour. Stopping.`, 'red');
    process.exit(1);
  }

  log(`🔄 Restarting backend (attempt ${restartCount}/${MAX_RESTART_ATTEMPTS})...`, 'yellow');

  if (backendProcess && !backendProcess.killed) {
    backendProcess.kill('SIGTERM');
  }

  setTimeout(() => {
    startBackend();
  }, RESTART_DELAY);
}

function performHealthCheck() {
  return new Promise((resolve) => {
    const port = process.env.PORT || 3000;
    const options = {
      hostname: 'localhost',
      port: port,
      path: '/api/health',
      method: 'GET',
      timeout: 10000
    };

    const req = http.request(options, (res) => {
      if (res.statusCode === 200) {
        resolve(true);
      } else {
        log(`⚠️  Health check failed with status: ${res.statusCode}`, 'yellow');
        resolve(false);
      }
    });

    req.on('error', (error) => {
      log(`⚠️  Health check error: ${error.message}`, 'yellow');
      resolve(false);
    });

    req.on('timeout', () => {
      log('⚠️  Health check timeout', 'yellow');
      req.destroy();
      resolve(false);
    });

    req.end();
  });
}

function startHealthCheck() {
  if (isShuttingDown) return;

  setInterval(async () => {
    if (isShuttingDown) return;

    log('🏥 Performing health check...', 'cyan');
    
    const isHealthy = await performHealthCheck();
    
    if (isHealthy) {
      log('✅ Health check passed', 'green');
    } else {
      log('❌ Health check failed - restarting backend', 'red');
      restartBackend();
    }
  }, HEALTH_CHECK_INTERVAL);

  log(`🏥 Health check started (interval: ${HEALTH_CHECK_INTERVAL}ms)`, 'blue');
}

// Keep-alive ping to prevent Render from sleeping
function startKeepAlivePing() {
  if (process.env.KEEP_ALIVE !== 'true') return;

  setInterval(() => {
    if (isShuttingDown) return;

    const port = process.env.PORT || 3000;
    http.get(`http://localhost:${port}/api/health`, (res) => {
      // Consume response to avoid memory leaks
      res.on('data', () => {});
      res.on('end', () => {
        log('📍 Keep-alive ping sent', 'cyan');
      });
    }).on('error', (error) => {
      log(`Keep-alive ping failed: ${error.message}`, 'yellow');
    });
  }, 840000); // Ping every 14 minutes (Render free tier sleeps after 15 minutes)
}

// Graceful shutdown handlers
function gracefulShutdown(signal) {
  log(`🛑 Received ${signal}. Starting graceful shutdown...`, 'yellow');
  isShuttingDown = true;

  if (backendProcess && !backendProcess.killed) {
    log('🛑 Stopping backend process...', 'yellow');
    backendProcess.kill('SIGTERM');
    
    // Force kill after 30 seconds
    setTimeout(() => {
      if (!backendProcess.killed) {
        log('💀 Force killing backend process...', 'red');
        backendProcess.kill('SIGKILL');
      }
    }, 30000);
  }

  setTimeout(() => {
    log('👋 Shutdown complete', 'green');
    process.exit(0);
  }, 5000);
}

// Signal handlers
process.on('SIGTERM', () => gracefulShutdown('SIGTERM'));
process.on('SIGINT', () => gracefulShutdown('SIGINT'));

// Unhandled rejection handler
process.on('unhandledRejection', (reason, promise) => {
  log(`Unhandled Rejection at: ${promise}, reason: ${reason}`, 'red');
});

// Start the application
log('🌟 KOL Tracker Backend Keep-Alive Wrapper Starting...', 'magenta');
log(`📊 Configuration:`, 'blue');
log(`   - Health Check Interval: ${HEALTH_CHECK_INTERVAL}ms`, 'blue');
log(`   - Max Restart Attempts: ${MAX_RESTART_ATTEMPTS}`, 'blue');
log(`   - Keep Alive: ${process.env.KEEP_ALIVE === 'true' ? 'Enabled' : 'Disabled'}`, 'blue');

startBackend();
startKeepAlivePing(); 

File: backend/telethon_service/keep-alive-wrapper.py
#!/usr/bin/env python3

"""
Keep-Alive Wrapper for KOL Tracker Telethon Service
Ensures the Telethon service stays running on Render with auto-restart on failures
"""

import asyncio
import subprocess
import signal
import time
import os
import sys
import logging
import aiohttp
from datetime import datetime, timedelta

# Configuration
HEALTH_CHECK_INTERVAL = int(os.getenv('HEALTH_CHECK_INTERVAL', '300'))  # 5 minutes
RESTART_DELAY = 5  # 5 seconds
MAX_RESTART_ATTEMPTS = 10
RESTART_WINDOW = 3600  # 1 hour

restart_count = 0
last_restart_time = 0
telethon_process = None
is_shutting_down = False

# Set up logging with colors
class ColoredFormatter(logging.Formatter):
    COLORS = {
        'DEBUG': '\033[36m',    # Cyan
        'INFO': '\033[32m',     # Green
        'WARNING': '\033[33m',  # Yellow
        'ERROR': '\033[31m',    # Red
        'CRITICAL': '\033[35m', # Magenta
    }
    RESET = '\033[0m'

    def format(self, record):
        log_color = self.COLORS.get(record.levelname, self.RESET)
        record.levelname = f"{log_color}{record.levelname}{self.RESET}"
        record.msg = f"{log_color}{record.msg}{self.RESET}"
        return super().format(record)

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='[%(asctime)s] %(levelname)s: %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S'
)
logger = logging.getLogger(__name__)

# Apply colored formatter
handler = logging.StreamHandler()
handler.setFormatter(ColoredFormatter('[%(asctime)s] %(levelname)s: %(message)s'))
logger.handlers = [handler]

async def start_telethon():
    """Start the Telethon service"""
    global telethon_process, is_shutting_down
    
    if is_shutting_down:
        return

    logger.info("🚀 Starting KOL Tracker Telethon Service...")

    try:
        telethon_process = await asyncio.create_subprocess_exec(
            'python', 'main.py',
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.STDOUT,
            env=os.environ.copy()
        )

        # Monitor process output
        asyncio.create_task(monitor_process_output())
        
        # Wait for process and handle exit
        exit_code = await telethon_process.wait()
        
        if is_shutting_down:
            logger.info("🛑 Telethon shutdown complete")
            return

        if exit_code == 0:
            logger.info("✅ Telethon exited gracefully")
        else:
            logger.error(f"💥 Telethon exited with code {exit_code}")
            await restart_telethon()

    except Exception as error:
        logger.error(f"❌ Telethon process error: {error}")
        await restart_telethon()

async def monitor_process_output():
    """Monitor and log Telethon process output"""
    global telethon_process
    
    if not telethon_process or not telethon_process.stdout:
        return

    try:
        async for line in telethon_process.stdout:
            line_str = line.decode('utf-8').strip()
            if line_str:
                # Forward Telethon logs with prefix
                logger.info(f"📡 Telethon: {line_str}")
    except Exception as e:
        logger.warning(f"Error reading process output: {e}")

async def restart_telethon():
    """Restart the Telethon service with backoff and limits"""
    global restart_count, last_restart_time, telethon_process, is_shutting_down

    if is_shutting_down:
        return

    current_time = time.time()
    
    # Reset restart count if enough time has passed
    if current_time - last_restart_time > RESTART_WINDOW:
        restart_count = 0

    restart_count += 1
    last_restart_time = current_time

    if restart_count > MAX_RESTART_ATTEMPTS:
        logger.error(f"🚨 Maximum restart attempts ({MAX_RESTART_ATTEMPTS}) reached within 1 hour. Stopping.")
        sys.exit(1)

    logger.warning(f"🔄 Restarting Telethon (attempt {restart_count}/{MAX_RESTART_ATTEMPTS})...")

    # Kill existing process if still running
    if telethon_process and telethon_process.returncode is None:
        try:
            telethon_process.terminate()
            await asyncio.wait_for(telethon_process.wait(), timeout=10)
        except asyncio.TimeoutError:
            logger.warning("💀 Force killing Telethon process...")
            telethon_process.kill()
            await telethon_process.wait()

    # Wait before restarting
    await asyncio.sleep(RESTART_DELAY)
    
    # Start new process
    await start_telethon()

async def perform_health_check():
    """Perform health check on Telethon service"""
    try:
        port = os.getenv('PORT', '8000')
        url = f"http://localhost:{port}/health"
        
        timeout = aiohttp.ClientTimeout(total=10)
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.get(url) as response:
                if response.status == 200:
                    return True
                else:
                    logger.warning(f"⚠️  Health check failed with status: {response.status}")
                    return False
                    
    except Exception as error:
        logger.warning(f"⚠️  Health check error: {error}")
        return False

async def health_check_loop():
    """Run periodic health checks"""
    logger.info(f"🏥 Health check started (interval: {HEALTH_CHECK_INTERVAL}s)")
    
    # Wait for service to start up
    await asyncio.sleep(30)
    
    while not is_shutting_down:
        logger.info("🏥 Performing health check...")
        
        is_healthy = await perform_health_check()
        
        if is_healthy:
            logger.info("✅ Health check passed")
        else:
            logger.error("❌ Health check failed - restarting Telethon")
            await restart_telethon()
            break  # Exit loop as restart will create new health check
            
        await asyncio.sleep(HEALTH_CHECK_INTERVAL)

async def keep_alive_ping():
    """Send keep-alive pings to prevent service sleeping"""
    if os.getenv('KEEP_ALIVE') != 'true':
        return

    logger.info("📍 Keep-alive ping started (14 minute intervals)")
    
    while not is_shutting_down:
        await asyncio.sleep(840)  # 14 minutes
        
        if is_shutting_down:
            break
            
        try:
            port = os.getenv('PORT', '8000')
            url = f"http://localhost:{port}/health"
            
            timeout = aiohttp.ClientTimeout(total=5)
            async with aiohttp.ClientSession(timeout=timeout) as session:
                async with session.get(url) as response:
                    logger.info("📍 Keep-alive ping sent")
        except Exception as error:
            logger.warning(f"Keep-alive ping failed: {error}")

async def graceful_shutdown(signum):
    """Handle graceful shutdown"""
    global is_shutting_down, telethon_process
    
    logger.warning(f"🛑 Received signal {signum}. Starting graceful shutdown...")
    is_shutting_down = True

    if telethon_process and telethon_process.returncode is None:
        logger.warning("🛑 Stopping Telethon process...")
        try:
            telethon_process.terminate()
            await asyncio.wait_for(telethon_process.wait(), timeout=30)
        except asyncio.TimeoutError:
            logger.warning("💀 Force killing Telethon process...")
            telethon_process.kill()
            await telethon_process.wait()

    logger.info("👋 Shutdown complete")
    sys.exit(0)

def signal_handler(signum, frame):
    """Signal handler for graceful shutdown"""
    if not is_shutting_down:
        asyncio.create_task(graceful_shutdown(signum))

async def main():
    """Main function"""
    logger.info("🌟 KOL Tracker Telethon Keep-Alive Wrapper Starting...")
    logger.info("📊 Configuration:")
    logger.info(f"   - Health Check Interval: {HEALTH_CHECK_INTERVAL}s")
    logger.info(f"   - Max Restart Attempts: {MAX_RESTART_ATTEMPTS}")
    logger.info(f"   - Keep Alive: {'Enabled' if os.getenv('KEEP_ALIVE') == 'true' else 'Disabled'}")

    # Set up signal handlers
    signal.signal(signal.SIGTERM, signal_handler)
    signal.signal(signal.SIGINT, signal_handler)

    # Start services concurrently
    tasks = [
        asyncio.create_task(start_telethon()),
        asyncio.create_task(health_check_loop()),
        asyncio.create_task(keep_alive_ping())
    ]

    try:
        await asyncio.gather(*tasks)
    except KeyboardInterrupt:
        await graceful_shutdown('SIGINT')
    except Exception as e:
        logger.error(f"Unexpected error in main loop: {e}")
        await graceful_shutdown('ERROR')

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        logger.info("👋 Interrupted by user")
    except Exception as e:
        logger.error(f"Fatal error: {e}")
        sys.exit(1) 